### PR TITLE
[TRIBAL 5.4] Implement tribal_set_context handler

### LINKED ISSUE #76: Implement tribal_set_context handler
## Summary

Implement the `tribal_set_context` handler to apply partial updates to the per-connection session state (project, model, provider) and return the full session context after mutation.

## Problem / Motivation

The MCP framework (5.1–5.3) established handler stubs, session state, and the mapping layer, but all seven tool handlers currently return a `FailedPrecondition` stub. `tribal_set_context` is the sole mutation path for session state (Tools §3.2, mcp_server_handler §6). Without it, agents cannot declare their model identity, override the resolved project, or confirm the session state — all of which are prerequisites for meaningful use of every other tool.

### In Scope
- [ ] Add a `PgPool` field to `TribalServerHandler` and update its constructor to accept one
- [ ] Implement `handle_set_context` in `@crates/tribal-mcp/src/handlers/set_context.rs`
- [ ] Parse `McpSetContextRequest` from the raw `serde_json::Value` params
- [ ] Validate `project_id` prefix and UUID structure via `ProjectId::from_str`
- [ ] Resolve validated `project_id` against the database via `ProjectRepository::find_by_id`
- [ ] Apply partial update to `SessionContext` (only fields present in the request are changed)
- [ ] Emit `notifications/resources/updated` for `tribal://session/context` via `notify_session_updated`
- [ ] Return `McpSetContextResponse` with the full post-mutation session state
- [ ] Unit tests covering happy path, error paths, and edge cases

### Out of Scope
- Two-tier scope policy splitting `tribal:read` (identity fields) from `tribal:write` (project override) — noted in Tools §8.6 as a likely revision once the scope system is built
- Auto-resolution of project from git remote at connection time — that is an M6 startup concern
- Field clearing semantics — the input schema defines all fields as `"type": "string"`, so `null` is not a valid input value; fields can only be set, not cleared
- Server-side enforcement of `additionalProperties: false` — the input schema forbids unknown fields, but this is a client-side contract; `McpSetContextRequest` uses default serde deserialisation which silently ignores unknown fields, consistent with how MCP servers typically handle schema validation

## Design Constraints

- `tribal_set_context` is the only mutation path for session state in M5. No other handler may write to `SessionContext` (mcp_server_handler §6).
- The handler must not perform silent inference from ambient context not present in the request. All resolution steps are explicit and visible in the handler implementation (mcp_server_handler §6).
- This tool does not write to the database. It mutates in-memory session state for the lifetime of the connection. The only database interaction is a read to resolve `project_id` (Tools §3.2).
- All fields are optional. The tool applies a partial update — only fields present in the request are changed (Tools §3.2).
- `model` and `provider` are free-form strings with no validation — the set of possible values is not bounded (Tools §3.2).
- `project_id` requires a valid `proj_`-prefixed ID for a project that exists in the database. Setting a non-existent project returns `not_found` (Tools §3.2).
- The response returns the full session context after the update (Tools §3.2).
- The tool is idempotent — setting the same values multiple times has no side effects (Tools §3.2).
- Never hold the session `RwLock` guard across an await point. Extract primitives under the lock, drop the guard, then proceed with async work (mcp_server_handler §5).
- Handler responsibilities follow the contract: parse request, extract session primitives (drop lock), call domain layer, map response, map errors at handler boundary only (mcp_server_handler §5).
- Domain and repository layers return domain errors. MCP error mapping happens only at the handler boundary via `IntoMcpError` (mcp_server_handler §5).
- Malformed JSON params produce a JSON-RPC `INVALID_PARAMS` protocol error via the `invalid_argument` helper in `@crates/tribal-mcp/src/error.rs`. This is distinct from the application-level `invalid_argument` error code (Tools §2.2) used for ID prefix/format failures, which produces an `isError: true` tool result via `IntoMcpError`.
- ID prefix validation uses `ProjectId::from_str` which returns `IdParseError` on wrong prefix or invalid UUID (Tools §2.1).
- Structured content convention: response includes both a human-readable text content block and structured JSON via `IntoCallToolResult` (Tools §2.3, §2.7).
- Auth is enforced at dispatch level in `call_tool`, not in this handler (mcp_server_handler §5). The current dispatch-level check uses a single `tribal:write` gate. Do not add handler-level scope checks — the intended two-tier split (Tools §8.6) will be implemented when the scope system is built.
- `RequestContext<RoleServer>` is present in the handler signature uniformly — used here for `notify_session_updated` via `context.peer` (mcp_server_handler §5).
- Resource update notifications are fire-and-forget — notification failure must not affect tool success or failure semantics (mcp_server_handler §6).
- Doc comments are required on all public types and functions (Conventions §14).

## Acceptance Criteria

- [ ] Empty request `{}` succeeds and returns the unchanged session context (no fields modified)
- [ ] Setting `model` and `provider` updates `SessionActor` fields on the session and returns them in the response
- [ ] Setting `project_id` with a valid prefixed ID for an existing project updates `SessionProject` on the session
- [ ] Response `project` field contains the resolved project's prefixed `id` (`proj_...`), `name`, and `git_remote` after a successful project override
- [ ] When no project is set, the response `project` field is `null` (not absent) — consistent with the existing `McpSetContextResponse` serialisation established in 5.3
- [ ] Setting `project_id` with a valid prefixed ID for a non-existent project returns a `not_found` application error
- [ ] Setting `project_id` with an invalid prefix (e.g. `ki_...`) returns an application-level `invalid_argument` error
- [ ] Setting `project_id` with an unparseable UUID suffix returns an application-level `invalid_argument` error
- [ ] Partial updates are additive — setting `model` does not clear a previously set `provider`
- [ ] Calling the handler twice with the same values produces the same response (idempotent)
- [ ] Response always contains `principal_key` and the full `actor` shape (`client_name`, `client_version`, `model`, `provider`), even for fields not settable via this tool
- [ ] The session `RwLock` guard is never held across an await point
- [ ] `notify_session_updated` is called after successful mutation; notification failure does not cause the handler to return an error
- [ ] Response uses the structured content convention (text block + `structured_content` JSON)
- [ ] No new clippy warnings with `--all-targets`

## Technical Notes

`TribalServerHandler` currently has no pool reference — it holds only repositories and session. Repository methods require `&mut PgConnection` as an explicit parameter, so the handler needs a `PgPool` to acquire connections. Add a `pool: PgPool` field to `TribalServerHandler` and accept it in the constructor. When `AppState` is introduced in M6, the constructor will receive `app_state.pool_mcp`. Existing tests must be updated to pass a pool from `test_context()`.

The handler flow is: parse `McpSetContextRequest` from params, then branch on whether `project_id` is present. If present, parse the prefixed string to `ProjectId` via `FromStr`, acquire a connection from the pool, and call `ProjectRepository::find_by_id`. If the project exists, build a `SessionProject` from the returned `Project` domain type. Then take a write lock on the session, apply all present fields (project, model, provider), build `McpSetContextResponse` from the updated session, and drop the lock. Finally call `notify_session_updated` and return the response.

The locking order matters: the project lookup is async (DB call), so it must happen before acquiring the session write lock. The session mutation itself is synchronous (field assignment), so the write lock can be acquired and released without crossing an await point.

`ProjectRepository::find_by_id` returns `Result<Project, DbError>` — a missing project produces `DbError::NotFound`, not `None`. This `DbError::NotFound` maps to a `not_found` MCP application error via the `IntoMcpError` implementation in `@crates/tribal-mcp/src/mapping/error.rs`. Note: Tools §2.2 lists "non-existent project_id" under `invalid_argument`, but Tools §3.2 explicitly specifies `not_found` for this tool. Follow §3.2 — the `DbError::NotFound` → `not_found` mapping already implements this.

All existing mapping types (`McpSetContextRequest`, `McpSetContextResponse`, `McpSessionProject`, `McpSessionActor`) and the `IntoCallToolResult` implementation for `McpSetContextResponse` are already defined in `@crates/tribal-mcp/src/mapping/session.rs`. The `notify_session_updated` helper in `@crates/tribal-mcp/src/session.rs` takes `&RwLock<SessionContext>` and `&Peer<RoleServer>` as arguments, acquires a read lock to check subscription status, and wraps the `peer.notify_resource_updated` call. Handlers should call this helper rather than making the peer notification directly. The `invalid_argument` error constructor is in `@crates/tribal-mcp/src/error.rs`.

The resource subscription lifecycle (list_resources, read_resource, subscribe, unsubscribe) is fully implemented in 5.1–5.2, and the server advertises resource subscription support in its capabilities. The `notify_session_updated` call in this handler completes the notification side of that lifecycle for session state changes.

Handler tests use mock repositories but still require `test_context()` for the connection parameter — the mock ignores the connection but the trait signature is mandatory. Configure the mock to return a project for happy-path tests and to return a not-found error for missing-project tests (since `find_by_id` signals absence via `DbError::NotFound`, not `Option`).

### PR DESCRIPTION
**Pull request by samfolo**
Implements the `tribal_set_context` handler.

Makes `TribalServerHandler` pool-aware.

Does a sweep of file formatting.

Closes #76.

### CODE REVIEW

File: .claude/settings.json [hunk 1]
```diff
@@ -60,6 +60,12 @@
 
       "Bash(grep:*)",
       "Bash(ls:*)"
+    ],
+    "deny": [
+      "Read(.env)",
+      "Read(.env.*)",
+      "Bash(git push --force*)",
+      "Bash(git push -f*)"
```
Comment (samfolo) [inline .claude/settings.json]: Decent tip I saw. Nothing happened, dw.

File: crates/tribal-mcp/src/handlers/set_context.rs [hunk 2]
```diff
@@ -1,26 +1,487 @@
...
+                Ok(id) => id,
+                Err(e) => return Ok(e.into_mcp_error().into_call_tool_result()),
+            };
+
+            let Ok(mut conn) = pool.acquire().await else {
+                return Ok(DbError::PoolExhausted {
+                    pool_name: POOL_NAME,
+                }
+                .into_mcp_error()
+                .into_call_tool_result());
```
[the remark this answers is not in the document]
Comment (samfolo) [reply crates/tribal-mcp/src/handlers/set_context.rs]: Great catch; always point out sub-standard error handling like this!

File: crates/tribal-mcp/src/server_handler.rs [hunk 3]
```diff
@@ -256,15 +270,13 @@ mod tests {
...
-    // -----------------------------------------------------------------------
-    // get_info tests
-    // -----------------------------------------------------------------------
+    // -- get_info -----------------------------------------------------------
 
-    #[test]
-    fn test_get_info_advertises_resources() {
+    #[tokio::test]
+    async fn test_get_info_advertises_resources() {
         let handler = test_handler();
```
[the remark this answers is not in the document]
Comment (samfolo) [reply crates/tribal-mcp/src/server_handler.rs]: CI is green and the lazy pools need async for construction. Not an issue.

File: crates/tribal-mcp/src/handlers/set_context.rs [hunk 4]
```diff
@@ -1,26 +1,531 @@
...
-        Ok(McpToolError {
-            code: McpErrorCode::FailedPrecondition,
-            message: "tribal_set_context is not yet implemented".into(),
-            details: serde_json::json!({}),
+        let (result, mutated) =
+            Self::apply_set_context(&self.pool, &self.repositories, &self.session, params).await?;
+
+        if mutated {
+            notify_session_updated(&self.session, &context.peer).await;
+        }
```
[the remark this answers is not in the document]
Comment (samfolo) [reply crates/tribal-mcp/src/handlers/set_context.rs]: Can't test this.
